### ISSUE TITLE
Don't use should

### ISSUE THREAD

**Issue by andreareginato** [andreareginato/betterspecs#15]
Write your thoughts about the "[don't use should](http://betterspecs.org/#should)" best practice.

**Comment by hellboy81** [comment]
Why should I **not** use  should syntax (i.e. `should.js` in _JavaScript_) ?

**Comment by baweaver** [comment]
That one more of got shoehorned into this. The original point was that using one liners only saves you space, and shouldn't be used frequently.

*<OPINION>*

On to the second point, and this is highly personal in preference as to using ``should`` in description strings. I don't like it quite simply because you're not certain. It should do something? It either does or it doesn't, and using flimsy wording doesn't make a concrete description. Granted that's soapboxing, but I still dislike seeing tests that lack a read of certainty to them. You need to be confident in your specs, not timid.

As an example:

```ruby
# Bad - Timid, and a fallback to old syntax
it 'should return 3 when given 1' do
  expect(method(1)).to eq(3)
end

# Good - Actionable, strong verb describing the test.
it 'returns 3 when given 1' do
  expect(method(1)).to eq(3)
end
```

*</OPINION>*

@hellboy81 - Mate, hate to break it to you but this is RSPEC, not Jasmine / Should.js. General concepts apply, but definitely the wrong area to debate this one.

**Comment by andypalmer** [comment]
Dan deliberately chose "should" when creating BDD to introduce the element of uncertainty; specifically with regards to our understanding of the system we are specifying.
"It should return 3 when given 1" is effectively short-hand for "based on our current understanding and assumptions, we expect method(1) to return 3. If, at some point in the future, this is not the case, we should be prepared to accept that our understanding was perhaps incomplete and that this expectation is no longer true"

The problem with the assertive case is that it invites no (explicit) doubt. If it fails, then the implementation must be broken.

It's a small distinction; and competent developers will implicitly question their assumptions, but Dan created BDD as a language-shift to help bring those whose understanding was not in that place up to speed in an simpler way. This is why the best TDDers see no difference between TDD and BDD and why beginners write better specifications when given BDD thinking tools

**Comment by Yvem** [comment]
@andypalmer +1

**Comment by tgaff** [comment]
I agree with audionerd.

This specification is about the test **description** the code inside should be the same.  There's already a separate entry for should vs. expect syntax.

**Comment by anulaibar** [comment]
Must agree with @baweaver. The tests are the blueprint for the code under test and therefore *should* does not belong in test descriptions.
To quote [Wikipedia](https://en.wikipedia.org/wiki/Unit_testing#Benefits):

> A unit test provides a strict, written contract that the piece of code must satisfy

The word _must_ is key here. Test descriptions serve as documentation of the code and should be clear and precise without room for interpretation. There is no _maybe_. Either the code does it or it doesn't do it.

To quote Yoda:
> Do. Or do not. There is no should.

**Comment by ro31337** [comment]
`should` is more searchable IMO, it doesn't modify the verb. Consider:

```ruby
class World
  def answer
    return 42
  end
end

# Good, when I search "return 42" I get two results: 1 in the code, 1 in test
it 'should return 42' do
  expect(world.answer).to eq(42)
end

# Bad, when I search "return 42" I get one result: in the code
it 'returns 42' do
  expect(world.answer).to eq(42)
end
```

I know that I can be smart and always use regex to search my codebase. But this is what I don't do.

**Comment by anulaibar** [comment]
The case when the return value is searchable isn't that common I would say. Consider this example:
```
class World
  def answer(a)
    return a
  end
end

it 'should return 42' do
  expect(world.answer(42)).to eq(42)
end
```
How would you search for that and expect a result where both the test and the code shows up?

A better approach is to keep your test in the same directory as the code it's testing so that you don't need to search for it.

**Comment by ywen** [comment]
Agreed with @anulaibar. And even if it is retuning `42`, one would write the code as

```ruby
def answer
  42
end
```

and 

```ruby
it 'returns 42' do
  expect(world.answer).to eq(42)
end
```

Now even you must use `return 42` in the production code. When you search for `return 42` it will simply return the prod code because the test says `returns 42` with a "s". When you search for `returns 42` then it will show the test code, not the prod code. So I don't think this is an issue

**Comment by ro31337** [comment]
yea, maybe it's a bad example. But it's nice when verb is unmodified IMHO.

**Comment by kilathaar** [comment]
"Should" is an excellent word to use in tests/examples because it encourages the reader of the test to question whether the test actually does what it say and whether the test should exist at all (refactoring).

In my opinion it is a sound attitude to always question tests hence always include "should".

**Comment by eduardomoroni** [comment]
```rb
def answer
  42
end

it 'returns 42' do
  expect(world.answer).to eq(42)
end
```

My questioning is, does the test ensure that: 
Is your code returning 42?
or,
Should your code return 42?

If you say that your test ensures that `your code is returning 42` and the test breaks, then your test is wrong. Because it's saying that your code is returning something that is not returning.
What if your test ensures that `your code should return 42` and your test breaks? Then or your test is wrong or your code is wrong. Because your test is NOT ensuring what your code is returning, it's ensuring what your code should return and somehow is not returning. So if you trust on your test you should look why your code is not returning 42.

**Comment by marnen** [comment]
@abelvarga I’d say both represent negation equally well. (I now use the latter form, but for completely unrelated reasons.)

**Comment by rangeoshun** [comment]
@marnen, I mean the difference is a lot more significant.

**Comment by anulaibar** [comment]
Love it that this discussion is still alive after six years ❤️ 
@marnen I'm curious, what made you change your mind?

**Comment by marnen** [comment]
@anulaibar Mostly that common practice shifted and I didn’t really feel strongly enough about keeping “should” that I was willing to buck the trend. Also that the tests (if they pass) describe what it *does*. (I still use “should” in my Cucumber stories, FWIW.)

**Comment by rangeoshun** [comment]
The change has started:

https://www.rubydoc.info/gems/rubocop-rspec/1.2.1/RuboCop/Cop/RSpec/ExampleWording

**Comment by lekhasy** [comment]
If you can't convince the world to think like you, why don't you just give them freedom by default?

Personally, I use the "should" word a lot, not only in TDD, BDD, but in daily communication with my client, with my colleague. 
The word "should" makes our brain think differently, it think about what the system gonna do for us, not how, and that lead to better tests.
And this is Ruby, right? Why don't you support I write test the way I speak?
Respect the different guys, and give half of the world the freedom they need.

**Comment by Potherca** [comment]
Being a non-ruby dev, coming from a background where definitions of words are heavily specified (for instance [RFC 2119](https://datatracker.ietf.org/doc/html/rfc2119) MUST / SHALL / SHOULD / MAY) and _the same words_ being used in both code and business domains, can someone tell me if this should vs. no-should thing is more about the wording or the (Rspec) syntax?

**Comment by akz92** [comment]
Hi @Potherca, in my understanding it is not a difference in meaning, it is more of an (IMO) improvement in RSpec's syntax and consequently in wording to make tests easier to read. This is somewhat frequent in the Ruby ecosystem as the language itself tries to make it possible to write code that read similar to how you'd write it in plain English.

**Comment by Potherca** [comment]
Ah, cool. Thanks for the information!
